### PR TITLE
feat: add global search for events, experiments and archives.

### PR DESCRIPTION
**Pull request by wanyoulc**
### What problem does this PR solve?

This PR adds a search trigger button, and a global search modal for searching events, experiments, and archives more conveniently.

### What is changed and how does it work?
-  update /ui/lib/search to add search logic, which use automata to achieve keyword matching
-  add a query parameter named event_id in EventsTable for linking to appropriate event when a user clicks search item
-  add a component named SearchTrigger, which is a search trigger button on Header
-  add a component named Search, which contains the main part of the search functionality

### Checklist

Tests

- [ ] Unit test
- [ ] E2E test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Breaking backward compatibility

Related changes

- [ ] Need to update the documentation

### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

**Comment by fewdan** [comment]
@wanyoulc    Please sign off DCO first.

File: ui/src/components/EventDetail/index.tsx [hunk 1]
```diff
@@ -16,85 +18,109 @@ const TableCell = withStyles({
...
-const EventDetail: React.FC<EventDetailProps> = ({ event: e }) => {
+const EventDetail: React.FC<EventDetailProps> = ({ eventID }) => {
   const { lang } = useSelector((state: RootState) => state.settings)
 
   const runningLabel = useRunningLabelStyles()
 
+  const [loading, setLoading] = useState(false)
+  const [e, setEvent] = useState<Event | undefined>(undefined)
+
+  const fetchEvent = () => {
```
Comment (g1eny0ung) [inline ui/src/components/EventDetail/index.tsx:32]: @fewdan Seems that we need to create an interface to get individual events.

File: ui/src/components/TopContainer/Header.tsx [hunk 2]
```diff
@@ -51,13 +58,69 @@ const useStyles = makeStyles((theme: Theme) =>
...
+      borderRadius: '4px',
+      color: '#969faf',
+      cursor: 'pointer',
+      '&:hover': {
+        color: '#1c1e21',
+      },
+    },
+    searchModal: {
+      position: 'relative',
+      maxWidth: '35rem',
```
Comment (g1eny0ung) [inline ui/src/components/TopContainer/Header.tsx:71]: I think it's better to use the percent to define the width, px for height.

File: ui/src/components/TopContainer/Header.tsx [hunk 3]
```diff
@@ -51,13 +58,69 @@ const useStyles = makeStyles((theme: Theme) =>
...
+      outline: 0,
+    },
   })
 )
 
 function hasLocalBreadcrumb(b: string) {
   return ['overview', 'experiments', 'newExperiment', 'events', 'archives', 'settings'].includes(b)
 }
 
+const SearchTrigger: React.FC = () => {
```
Comment (g1eny0ung) [inline ui/src/components/TopContainer/Header.tsx:85]: Encapsulate this component into a file?

Also, add a global shortcut to trigger searching? like `Ctrl/CMD + P`.

**Review by g1eny0ung — CHANGES_REQUESTED** [review]

**Comment by g1eny0ung** [comment]
@wanyoulc And also, the new searching styles don't fit the dark mode. PTAL.

**Review by g1eny0ung — APPROVED** [review]
LGTM

**Review by fewdan — DISMISSED** [review]
LGTM

**Comment by fewdan** [comment]
Please resolve the conflict files.  @wanyoulc

**Review by g1eny0ung — APPROVED** [review]

**Review by zhouqiang-cl — APPROVED** [review]
LGTM

**Pull request merged**